contracts/simple_dex/lib.rs

```diff
@@ -0,0 +1,660 @@
+        #[ink(topic)]
+        token_in: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct TokenToToken {
+        caller: AccountId,
+        #[ink(topic)]
+        token_in: AccountId,
+        #[ink(topic)]
+        token_out: AccountId,
+        amount_in: Balance,
+        amount_out: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct NativeDeposit {
+        caller: AccountId,
+        amount: Balance,
+        minted_liquidity: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct NativeWithdrawal {
+        caller: AccountId,
+        amount: Balance,
+        redeemed_liquidity: Balance,
+    }
+
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct SwapFeeSet {
+        #[ink(topic)]
+        caller: AccountId,
+        swap_fee: Balance,
+    }
+
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct SimpleDex {
+        /// total supply of pool shares
+        pub total_liquidity: u128,
+        /// tracks pool shares per account
+        pub liquidity: Mapping<AccountId, u128>,
+        pub swap_fee: u128,
+        pub access_control: AccountId,
+        pub ubik: AccountId,
+        pub cyberiad: AccountId,
+        pub lono: AccountId,
+    }
+
+    impl AccessControlled for SimpleDex {
+        type ContractError = DexError;
+    }
+
+    impl SimpleDex {
+        #[ink(constructor)]
+        pub fn new(ubik: AccountId, cyberiad: AccountId, lono: AccountId) -> Self {
+            let caller = Self::env().caller();
+            let code_hash = Self::env()
+                .own_code_hash()
+                .expect("Called new on a contract with no code hash");
+            let required_role = Role::Initializer(code_hash);
+            let access_control = AccountId::from(ACCESS_CONTROL_PUBKEY);
+
+            let role_check = <Self as AccessControlled>::check_role(
+                access_control,
+                caller,
+                required_role,
+                Self::cross_contract_call_error_handler,
+                Self::access_control_error_handler,
+            );
+
+            match role_check {
+                Ok(_) => {
+                    initialize_contract(|contract| Self::new_init(contract, ubik, cyberiad, lono))
+                }
+                Err(why) => panic!("Could not initialize the contract {:?}", why),
+            }
+        }
+
+        /// Swaps the transferred amount of native token to one of the pools PSP22 tokens
+        #[ink(message, payable)]
+        pub fn native_to_token(&mut self, token_out: AccountId) -> Result<(), DexError> {
+            let amount_token_in = self.env().transferred_value();
+            // we use the balance of the native token just before the exchange tx
+            let balance_token_in = self
+                .env()
+                .balance()
+                .checked_sub(amount_token_in)
```

Comment:
we are alredy given this `payable`, right? or is it somehow hangin

Reply:
not sure I follow this comment, could you elaborate pls?

Reply:
this `message` (method) will be accompanied by some amount of native token (since it is payable); my question is whether these funds will be already available in this place